### PR TITLE
feat(helm): import child values to parent

### LINKED ISSUE #1995: proposal: allow import of child values via optional requirements field 'import-values'
# Problem

When working on complicated nested charts developers would like to create 'library charts' that contain common values that need to be accessible from the parent chart but there is no mechanism to do this. Developers resort to workarounds such as build scripts that merge values files which add more steps and are not helm native.

# Proposed Solution

Allow the import of a child's values to parent values by adding a `imports-values:` field in requirements.yaml as in the example below

```
# requirements.yaml (simple list)
dependencies:
  - name: kolla-common
    repository: http://localhost:10191
    version: 0.1.0
    import-values:
      - foo.thing
```

The above would instruct helm to take the values found at path "foo.thing" in sub-chart "kolla-common" and place them at the root of the top parent's values. So that given a parent values of 

````
# parent values
mything: yay
````

and child values
````
# child values
foo:
  thing:
    bar: true
    baz: false
    other:
      - 1
      - 2
      - 3
````

The resulting parent values would be 

````
# result parent values
mything: yay
bar: true
baz: false
  other:
    - 1
    - 2
    - 3
````

## Advanced usage

Additionally the `import-values:` field would support a child/parent dictionary value that would allow reparenting the imported values into a specific path in the parent values as in the example below:

````
# requirements (child/parent)
...
import-values:
  - child: foo.thing
    parent: new.key
````

In the above example, the values from the child path are placed at the specified location 'new.key' in the parent values as below:

````
# result values
mything: yay
new:
  key:
    bar: true
    baz: false
    other:
      - 1
      - 2
      - 3
````

## Non-existent import values

If a `import-values` field contains child paths that do not exist, a warning will be issued.

## Mixed string and child/parent

Mixing the simple list of strings and child/parent dictionaries is supported and the below would be valid:

````
# requirements.yaml
...
import-values:
  - common
  - child: foo.thing
    parent: new.key
````

## Exports key convention

To facilitate the introspection of values intended to be consumed by a parent, it is suggested to follow a `exports:` key convention wherein those values are all contained in field named `exports:` as below:

````
# child/values.yaml

exports:
  foo: true
  bar: false
baz:
  pike: true
....
````

By following the above convention, tools will be able to better inspect and when necessary exclude the values intended for parent. Users that need the flexibility to ignore the convention are allowed to do so with the caveat that future analysis involving values intended for parent may be difficult or impossible.

## Importing from Child to Parent to GrandParent

By using the imports-values mechanism at each level in the hierarchy, values can be imported from a child chart to its parent and then imported into the grandparent (and so on). 

````
# parent requirements
imports-values:
  - child: data.foo
    parent: imported-from-child

# grandparent requirements
import-values:
  - child: imported-from-child
    parent: imported-from-child-via-parent


````

### PR DESCRIPTION
**Pull request by jascott1**
Implements a mechanism in requirements.yaml to allow the import
and re-parenting of value table from child chart.

Closes #1995

### CODE REVIEW

File: pkg/chartutil/requirements.go [hunk 1]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+							coalesceTables(b, vv.AsMap())
+						} else {
+
+							vm := pathToMap(nm["parent"], vv.AsMap())
+							coalesceTables(b, vm)
+						}
+					}
+				case string:
+					nm := make(map[string]string)
+					nm["child"] = riv.(string)
```
Comment (kfox1111) [inline pkg/chartutil/requirements.go]: I think this should be "exports." + riv.(string)

**Review by kfox1111 — COMMENTED** [review]
Looks really good. I'll see if I can test this at some point. One default setting though is off I think. commented inline.

**Comment by technosophos** [comment]
Can you dump some docs on here as well?

File: pkg/chartutil/requirements.go (hunk 1 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: Doing that would hardcode the 'convention'. I could alter the tests to demonstrate the convention tho.

**Comment by jascott1** [comment]
@technosophos any thoughts on breaking out into another doc 'advanced usage' and moving all the tags, conditions and local path etc there?

**Comment by technosophos** [comment]
@jascott1 Yeah, that sounds like a good idea.

File: pkg/chartutil/requirements.go (hunk 1 above)
Comment (kfox1111) [inline pkg/chartutil/requirements.go]: that if the user was going to follow the convention,
```yaml
import-values:
  - mything
```
would be
```yaml
exports:
  mything: ...
```
If you wanted to not follow the convention of pulling from exports, you could still use the parent/child format to get at it:
```yaml
import-values:
  - child: mything
  - parent: .
```
would be
```yaml
mything: ...
```

Comment (jascott1) [inline pkg/chartutil/requirements.go]: I somehow forgot/missed that `-mything` and `child: mything` would be different with the former using exports and the latter not. Although Im not crazy about it, I get it and will get the PR fixed up asap. Thanks : )

Comment (jascott1) [inline pkg/chartutil/requirements.go]: I perceive a problem or inconvenience with simple list form. I don't see it as a deal breaker but you cannot just use simple list to go from child->parent->grandparent because the imported value comes in at "." not in "exports." so the simple list import in grandparent cannot access parents values imported from child (with simple list). Just FYI in case you had plans to do child->parent->grandparent with just simple list.

Comment (kfox1111) [inline pkg/chartutil/requirements.go]: For the primary use case, I dont really see child->parent->grandparent happening directly I think? so, for the kolla-kubernetes case again, the idea is the having a library chart, where the library provides a set of macro's, and a set of grouped values that should be included in the chart consuming the macros. The shared lib's macro values are then pulled into the parent directly, and are available for setting in the grand parent via the grandparents <parentchartname>: ... section.

Am I missing a common use case?

Comment (jascott1) [inline pkg/chartutil/requirements.go]: I think its ok since the complex form can be used where the simple doesn't suffice. A grandparent will likely use the complex form anyway since it will likely need to reparent the import.

Comment (kfox1111) [inline pkg/chartutil/requirements.go]: Ah. Ok. That makes sense to me.

**Comment by technosophos** [comment]
@jascott1 In the franticness of the last week, I lost track of this. Is it ready for a review?

**Comment by kfox1111** [comment]
Looks good to me.

File: pkg/chartutil/requirements.go [hunk 2]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
 
 	return nil
 }
+
+// pathToMap creates a nested map given a YAML path in dot notation
+func pathToMap(path string, data map[string]interface{}) map[string]interface{} {
+	ap := strings.Split(path, ".")
+	n := []map[string]interface{}{}
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] You could use `make` here with a size the pre-allocate the backing array, given that it's always going to hold `len(ap)` entries.

**Comment by jascott1** [comment]
@technosophos Its ready except for the docs. Should I just add to existing doc and then break out into 'Advanced Usage' in another PR later or do it now?

File: pkg/chartutil/requirements.go [hunk 3]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
 
 	return nil
 }
+
+// pathToMap creates a nested map given a YAML path in dot notation
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] I'm new to this code base, but I'm accustomed to seeing Godoc comments punctuated as sentences.

File: pkg/chartutil/requirements.go [hunk 4]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+func pathToMap(path string, data map[string]interface{}) map[string]interface{} {
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] I found this function hard to follow, so I rewrote it to understand it better.  
  I interpret as pointing the last path component to `data`, and each preceding path component to the next, where each component is represented as a map with a single entry, the key being the path component, and the value being the map holding the successor key.
  Does the following achieve the same goal?
```go
func pathToMap(path string, data map[string]interface{}) map[string]interface{} {
	components := strings.Split(path, ".")
/* What should happen if the path is empty?
	if len(components) == 0 {
		return nil
	}
*/
	sort.Sort(sort.Reverse(sort.StringSlice(components)))
	m := data
	for _, c := range components {
		m = map[string]interface{}{c: m}
	}
	return m
}
```

File: pkg/chartutil/requirements.go [hunk 5]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+		}
+	}
+
+	return n[0]
+}
+
+// getParents returns a slice of parent charts in reverse order
+func getParents(c *chart.Chart, out []*chart.Chart) []*chart.Chart {
+	if len(out) == 0 {
+		out = []*chart.Chart{}
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Alternately, omit the appending step:  
```go
out = []*chart.Chart{c}
```

File: pkg/chartutil/requirements.go [hunk 6]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+			}
+			n[i][k] = n[z]
+		}
+	}
+
+	return n[0]
+}
+
+// getParents returns a slice of parent charts in reverse order
+func getParents(c *chart.Chart, out []*chart.Chart) []*chart.Chart {
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] This is topological sort of the dependency graph, right?

File: pkg/chartutil/requirements.go [hunk 7]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+	}
+
+	return out
+}
+
+// processImportValues merges values from child to parent based on ImportValues field
+func processImportValues(c *chart.Chart, v *chart.Config) error {
+	reqs, err := LoadRequirements(c)
+	if err != nil {
+		log.Printf("Warning: ImportValues cannot load requirements for %s", c.Metadata.Name)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Why log here and return nil as opposed to augmenting and returning an error?

File: pkg/chartutil/requirements.go [hunk 8]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+	if err != nil {
+		log.Fatalf("Error coalescing values for ImportValues %s", err)
+	}
+	nv := v.GetValues()
+	b := make(map[string]interface{})
+	for kk, v3 := range nv {
+		b[kk] = v3
+	}
+	for _, r := range reqs.Dependencies {
+		if len(r.ImportValues) > 0 {
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Alternately, save yourself some indentation:  
```go
if len(r.ImportValues) == 0 {
  continue
}
```

File: pkg/chartutil/requirements.go [hunk 9]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+			var outiv []interface{}
+			for _, riv := range r.ImportValues {
+				switch riv.(type) {
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] You can omit the the type assertions at lines 333 and 354 by using [a short variable declaration](https://golang.org/ref/spec#TypeSwitchStmt):  
```go
switch iv := riv.(type) {
case map[string]interface{}:
  // Use "iv," which is guaranteed to be a map[string]interface{}.
  // ...
case string:
  // Use "iv," which is guaranteed to be a string.
  // ...
}
```

File: pkg/chartutil/requirements.go [hunk 10]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+						s := r.Name + "." + nm["child"]
+						vv, err := cvals.Table(s)
+						if err != nil {
+							log.Printf("Warning: ImportValues missing table %v", err)
+							continue
+						}
+						if nm["parent"] == "." {
+							coalesceTables(b, vv.AsMap())
+						} else {
+
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Remove blank line.

File: pkg/chartutil/requirements.go [hunk 11]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+		b[kk] = v3
+	}
+	for _, r := range reqs.Dependencies {
+		if len(r.ImportValues) > 0 {
+			var outiv []interface{}
+			for _, riv := range r.ImportValues {
+				switch riv.(type) {
+				case map[string]interface{}:
+					if m, ok := riv.(map[string]interface{}); ok {
+						nm := make(map[string]string)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Alternately:  
```go
nm := map[string]string {
  "child": m["child"].(string),
  "parent": m["parent"].(string),
}
```

File: pkg/chartutil/requirements.go [hunk 12]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+		}
+	}
+
+	cv, err := coalesceValues(c, b)
+	if err != nil {
+		log.Fatalf("Error coalescing processed values for ImportValues %s", err)
+	}
+	y, err := yaml.Marshal(cv)
+	if err != nil {
+		log.Printf("Warning: ImportValues could not marshall %v", err)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Why not return an error here?

File: pkg/chartutil/requirements.go [hunk 13]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+				}
+			}
+			// set our formatted import values
+			r.ImportValues = outiv
+		}
+	}
+
+	cv, err := coalesceValues(c, b)
+	if err != nil {
+		log.Fatalf("Error coalescing processed values for ImportValues %s", err)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Why not return an error here?

File: pkg/chartutil/requirements.go [hunk 14]
```diff
@@ -266,3 +268,128 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+
+	return nil
+}
+
+// ProcessRequirementsImportValues imports specified chart values from child to parent
+func ProcessRequirementsImportValues(c *chart.Chart, v *chart.Config) error {
+	pc := getParents(c, nil)
+	for i := len(pc) - 1; i >= 0; i-- {
+		processImportValues(pc[i], v)
+
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Remove blank line.

**Review by seh — CHANGES_REQUESTED** [review]
I didn't review the tests yet.

File: pkg/chartutil/requirements.go (hunk 10 above)
Comment (technosophos) [inline pkg/chartutil/requirements.go]: We don't tend to be too picky on whitespace.

**Comment by technosophos** [comment]
@jascott1 Yeah, just some simple documentation in the existing docs is good for now.

File: pkg/chartutil/requirements.go (hunk 7 above)
Comment (technosophos) [inline pkg/chartutil/requirements.go]: Yeah, I agree on the error logging. I think we should probably fail the entire process step if we can't build the data.

File: pkg/chartutil/requirements.go (hunk 3 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: Fixed.

File: pkg/chartutil/requirements.go (hunk 4 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: @seh Thanks for the catch on the 0 length value. I have updated to return nil in that case.

Your interesting solution does not preserve the order of the path elements however ("c.a.b" comes back as map[a:map[b:map[c:map[]]]] ). I played with your approach some but didn't get a solution but please let me know if you come up with another elegant one.

File: pkg/chartutil/requirements.go (hunk 2 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: This suggestion actually caused problems believe it or not. I don't think its worth it.

File: pkg/chartutil/requirements.go (hunk 5 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: Nice, fixed.

File: pkg/chartutil/requirements.go (hunk 7 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: I returned the error. Let me know if its preferable to fail it here instead.

File: pkg/chartutil/requirements.go (hunk 8 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: I will pass on this one, its arguably more grok-able the way it is (imo).

File: pkg/chartutil/requirements.go (hunk 11 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: Fixed.

File: pkg/chartutil/requirements.go (hunk 13 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: I removed the logging and just returned err.

File: pkg/chartutil/requirements.go (hunk 12 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: Returned the err.

File: pkg/chartutil/requirements.go (hunk 2 above)
Comment (seh) [inline pkg/chartutil/requirements.go]: I suspect that's because you used the two-argument form of `make`, as opposed to the three-argument form, right? I meant the following:
```go
n := make([]map[string]interface{}, 0, len(ap))
```
If instead you use the two-argument ` make([]map[string]interface{}, len(ap))`, then you shouldn't use `append` later; instead, just assign to the slice elements by index, because it's already of the intended size.

File: pkg/chartutil/requirements.go (hunk 4 above)
Comment (seh) [inline pkg/chartutil/requirements.go]: I was looking for a way to reverse a slice, searched the standard library, noticed the symbol `Reverse` in the `sort` package, glanced at the usage example, and off I went, not noticing—believe it or not— that using `sort` does in fact _sort the input_, which was not my intention.

Given that:
```go
unc reverseStrings(s []string) {
	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
		s[i], s[j] = s[j], s[i]
	}
}

func pathToMap(path string, data map[string]interface{}) map[string]interface{} {
	components := strings.Split(path, ".")
/* What should happen if the path is empty?
	if len(components) == 0 {
		return nil
	}
*/
	reverseStrings(components)
	m := data
	fmt.Println("components:", components)
	for _, c := range components {
		m = map[string]interface{}{c: m}
	}
	return m
}
```

**Comment by jascott1** [comment]
I mistakenly let imported values win the merge over parent. Please hold off on the reviews until I refactor.

File: pkg/chartutil/requirements.go (hunk 4 above)
Comment (seh) [inline pkg/chartutil/requirements.go]: @jascott1, does this alternative yield the same outcome as your `pathToMap`?

**Comment by jascott1** [comment]
This should be ready for review now.

File: docs/charts.md [hunk 15]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
   * The `tags:` key in values must be a top level key. Globals and nested `tags:` tables
     are not currently supported.
 
+#### Importing Child Values via requirements.yaml
+
+In some cases it is desirable to allow a child chart's values to propagate to the parent chart and be 
+shared as common defaults. An additional benefit of using the `exports` format is that it will enable future 
+tooling to introspect user settable values.
```
Comment (seh) [inline docs/charts.md]: - [ ] Add a hyphen between "user" and "settable."

File: docs/charts.md [hunk 16]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+The keys containing the values to be imported can be specified in the parent chart's requirements.yaml using
```
Comment (seh) [inline docs/charts.md]: - [ ] Elaborate on what "requirements.yaml" means.  
  Set it in italic and use the word "file."  
  > the parent chart's _requirements.yaml_ file

File: docs/charts.md [hunk 17]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+#### Importing Child Values via requirements.yaml
+
+In some cases it is desirable to allow a child chart's values to propagate to the parent chart and be 
+shared as common defaults. An additional benefit of using the `exports` format is that it will enable future 
+tooling to introspect user settable values.
+
+The keys containing the values to be imported can be specified in the parent chart's requirements.yaml using 
+a YAML list. Each item in the list is a key which is imported from the child chart's `exports` field. 
+
+To import values not contained in the `exports` key, use the [child/parent](#using-the-child/parent-format) format.
```
Comment (seh) [inline docs/charts.md]: - [ ] Add a short sentence here to clarify why we haven't seen examples of these formats you're describing.  
  > Examples of both formats follow.

File: docs/charts.md [hunk 18]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+##### Using the exports format
+
+If a child chart's values.yaml contains an `exports` field at the root, it's contents may be imported
```
Comment (seh) [inline docs/charts.md]: - [ ] Clarify that "values.yaml" is a file.  
  > chart's _values.yaml_ file contains
 - [ ] s/it's/its/

File: docs/charts.md [hunk 19]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+a YAML list. Each item in the list is a key which is imported from the child chart's `exports` field. 
+
+To import values not contained in the `exports` key, use the [child/parent](#using-the-child/parent-format) format.
+
+##### Using the exports format
+
+If a child chart's values.yaml contains an `exports` field at the root, it's contents may be imported 
+directly into the parent's values by specifying the keys to import as in the example below:
+
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Four backticks? I thought Markdown (and GHFM) uses only three.  
  In any case, specify the language that follows:
  > \```yaml

File: docs/charts.md [hunk 20]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+# parent's requirements.yaml
+    ...
+    import-values:
+      - data
+````  
+````
+# child's values.yaml
+
```
Comment (seh) [inline docs/charts.md]: - [ ] Eliminate this blank line, or include one like in the example above around line 323.

File: docs/charts.md [hunk 21]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+If a child chart's values.yaml contains an `exports` field at the root, it's contents may be imported 
+directly into the parent's values by specifying the keys to import as in the example below:
+
+````
+# parent's requirements.yaml
+    ...
+    import-values:
+      - data
+````  
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 22]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+To import values not contained in the `exports` key, use the [child/parent](#using-the-child/parent-format) format.
+
+##### Using the exports format
+
+If a child chart's values.yaml contains an `exports` field at the root, it's contents may be imported 
+directly into the parent's values by specifying the keys to import as in the example below:
+
+````
+# parent's requirements.yaml
```
Comment (seh) [inline docs/charts.md]: - [ ] Mention that "requirements.yaml" is a file.  
  > requirements.yaml file

File: docs/charts.md [hunk 23]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+    ...
+    import-values:
+      - data
+````  
+````
+# child's values.yaml
```
Comment (seh) [inline docs/charts.md]: - [ ] Mention that "values.yaml" is a file.  
  > values.yaml file

File: docs/charts.md [hunk 24]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+  data:
+    myint: 99
+````  
+
+Since we are specifying the key `data` in our import list, Helm looks in the the `exports` field of the child 
+chart for `data` key and imports its contents. 
+
+The final parent values would contain our exported field:
+
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 25]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+# parent's values
+...
+myint: 99
+
+````
+
+Please note the parent key `data` is not contained in the parent's final values. If you need to specify the 
+parent key, use the 'child/parent' format.
```
Comment (seh) [inline docs/charts.md]: - [ ] Note that that's coming next.  
  > format, as described in the next section.

File: docs/charts.md [hunk 26]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+To access values that are not contained in the `exports` key of the child chart's values, you will need to 
+specify the source key of the values to be imported (`child`) and the destination path in the parent chart's 
+values (`parent`).
+
+The `import-values` in the example below instructs Helm to take any values found at `child:` path and copy them 
+to the parent's values at the path specified in `parent:`
+
+````
+# parent's requirements.yaml
```
Comment (seh) [inline docs/charts.md]: - [ ] Mention that "requirements.yaml" is a file.  
  > requirements.yaml file

File: docs/charts.md [hunk 27]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+        version: 0.1.0
+        ...
+        import-values:
+          - child: default.data
+            parent: myimports
+````
+In the above example, values found at `default.data` in the subchart1's values will be imported
+to the `myimports` key in the parent chart's values as detailed below: 
+
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 28]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+##### Using the child/parent format
+
+To access values that are not contained in the `exports` key of the child chart's values, you will need to 
+specify the source key of the values to be imported (`child`) and the destination path in the parent chart's 
+values (`parent`).
+
+The `import-values` in the example below instructs Helm to take any values found at `child:` path and copy them 
+to the parent's values at the path specified in `parent:`
+
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 29]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+# parent's requirements.yaml
+dependencies:
+      - name: subchart1
```
Comment (seh) [inline docs/charts.md]: - [ ] The indentation looks exaggerated here.  
  I expected this list to be indented either zero or two spaces from the 'd' in "dependencies."

File: docs/charts.md [hunk 30]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+# parent's values
+
+myimports:
+  myint: 0
+  mybool: false
+  mystring: "helm rocks!"
+  
+
+````
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 31]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+default:
+  data:
+    myint: 999
+    mybool: true
+    
+````
+The parent chart's resulting values would be:
+
+````
```
Comment (seh) [inline docs/charts.md]: - [ ] Indicate the language that follows.  
  > \```yaml

File: docs/charts.md [hunk 32]
```diff
@@ -302,6 +302,108 @@ helm install --set tags.front-end=true --set subchart2.enabled=false
...
+
+myimports:
+  myint: 0
+  mybool: false
+  mystring: "helm rocks!"
+  
+
+````
+````
+# subchart1's values.yaml
```
Comment (seh) [inline docs/charts.md]: - [ ] Mention that "values.yaml" is a file.  
  > values.yaml file

File: pkg/chartutil/requirements.go [hunk 33]
```diff
@@ -62,6 +62,8 @@ type Dependency struct {
 	Tags []string `json:"tags"`
 	// Enabled bool determines if chart should be loaded
 	Enabled bool `json:"enabled"`
+	// ImportValues holds the mapping of source values to parent key to be imported
+	ImportValues []interface{} `json:"import-values"`
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] I had a hard time understanding the field's type from the documentation.  
  Why is it not of type `map[string]string`?

File: pkg/chartutil/requirements.go [hunk 34]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
 
 	return nil
 }
+
+// pathToMap creates a nested map given a YAML path in dot notation.
+func pathToMap(path string, data map[string]interface{}) map[string]interface{} {
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] I still find this function's algorithm hard to follow, but I believe you that it works.

File: pkg/chartutil/requirements.go [hunk 35]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+		if len(ch.Dependencies) > 0 {
+			out = append(out, ch)
+			out = getParents(ch, out)
+		}
+	}
+
+	return out
+}
+
+// processImportValues merges values from child to parent based on ImportValues field.
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Elaborate where the "ImportValues" field comes from.  
  > based on the chart's dependencies' ImportValues fields.

File: pkg/chartutil/requirements.go [hunk 36]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+	if err != nil {
+		return err
+	}
+	// combine chart values and its dependencies' values
+	cvals, err := CoalesceValues(c, v)
+	if err != nil {
+		return err
+	}
+	nv := v.GetValues()
+	b := make(map[string]interface{})
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Specify an initial size for the map.  
  ```go
  make(map[string]interface{}, len(nv))
   ```

File: pkg/chartutil/requirements.go [hunk 37]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+					nm := map[string]string{
+						"child":  iv["child"].(string),
+						"parent": iv["parent"].(string),
+					}
+					outiv = append(outiv, nm)
+					s := r.Name + "." + nm["child"]
+					// get child table
+					vv, err := cvals.Table(s)
+					if err != nil {
+						log.Printf("Warning: ImportValues missing table %v", err)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Precede the error with a colon.
  ```go
  table: %v"
  ```

File: pkg/chartutil/requirements.go [hunk 38]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+					b = coalesceTables(cvals, vm)
+				case string:
+					nm := make(map[string]string)
+					nm["child"] = "exports." + iv
+					nm["parent"] = "."
+					outiv = append(outiv, nm)
+					s := r.Name + "." + nm["child"]
+					vm, err := cvals.Table(s)
+					if err != nil {
+						log.Printf("Warning: ImportValues missing table %v", err)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Precede the error with a colon.
  ```go
  table: %v"

File: pkg/chartutil/requirements.go [hunk 39]
```diff
@@ -266,3 +268,127 @@ func ProcessRequirementsEnabled(c *chart.Chart, v *chart.Config) error {
...
+					vv, err := cvals.Table(s)
+					if err != nil {
+						log.Printf("Warning: ImportValues missing table %v", err)
+						continue
+					}
+					// create value map from child to be merged into parent
+					vm := pathToMap(nm["parent"], vv.AsMap())
+					b = coalesceTables(cvals, vm)
+				case string:
+					nm := make(map[string]string)
```
Comment (seh) [inline pkg/chartutil/requirements.go]: - [ ] Consider using the same initialization style as line 342 above.  
  ```go
  nm := map[string]string{
    "child": "exports." + iv,
    "parent": ".",
  }
  ```

File: pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml [hunk 40]
```diff
@@ -1,21 +1,37 @@
...
+  SCBint: 33
+  SCBstring: "boba"
+
+exports:
+  SCBexported1:
+    SCBexported1A:
+      SCBexported1B: 1965
+
+  SCBexported2:
+      SCBexported2A: "blaster"
```
Comment (seh) [inline pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml]: - [ ] The indentation from here down looks odd.  
  Above here, we're stepping by two spaces, but here we start stepping by four.

File: pkg/chartutil/testdata/subpop/charts/subchart1/values.yaml [hunk 41]
```diff
@@ -1,21 +1,55 @@
...
+  SCBstring: "murkwood"
+  SCBextra1: 13
+
+  SC1extra6: 77
+
+SCBexported1A:
+  SC1extra7: true
+
+exports:
+    SC1exported1:
```
Comment (seh) [inline pkg/chartutil/testdata/subpop/charts/subchart1/values.yaml]: - [ ] The indentation from here down looks odd.  
  Above here, we're stepping by two spaces, but here we start stepping by four.

**Review by seh — CHANGES_REQUESTED** [review]
I didn't review the unit test too carefully. I was mostly focusing on the documentation.

File: pkg/chartutil/requirements.go (hunk 33 above)
Comment (jascott1) [inline pkg/chartutil/requirements.go]: This is a list or sublist depending on the format. I updated the comment to explain more.

File: pkg/chartutil/requirements.go [hunk 42]
```diff
@@ -62,6 +62,9 @@ type Dependency struct {
 	Tags []string `json:"tags"`
 	// Enabled bool determines if chart should be loaded
 	Enabled bool `json:"enabled"`
+	// ImportValues holds the mapping of source values to parent key to be imported. Each item can be a
+	// string or pair of child/parent sublist items.
```
Comment (seh) [inline pkg/chartutil/requirements.go]: Thank you for clarifying that.

File: pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml (hunk 40 above)
Comment (seh) [inline pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml]: That line 22 still looks like it's indented too far, but I won't hold up the whole PR for it.

**Review by seh — APPROVED** [review]
I noted one remaining nit, but I'm excited to get this in!

File: pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml (hunk 40 above)
Comment (jascott1) [inline pkg/chartutil/testdata/subpop/charts/subchart1/charts/subchartB/values.yaml]: Fixed!

**Review by technosophos — APPROVED** [review]
This appears to be all ready to merge. Thank you very much @seh for the thorough review.
